### PR TITLE
Fix identifiers behavior bugs

### PR DESCRIPTION
**Pull request by soyuka**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #551 
| License       | MIT
| Doc PR        | ∅

#550 ref (sorry about multiple PR, github was buggy).

### CODE REVIEW

File: src/Bridge/Symfony/Routing/IriConverter.php [hunk 1]
```diff
@@ -77,25 +77,72 @@ public function getIriFromItem($item, int $referenceType = UrlGeneratorInterface
...
+            return rawurlencode($identifiers);
+        }
+
+        if (1 === count($identifiers)) {
+            return rawurlencode(array_values($identifiers)[0]);
+        }
+
+        return array_map(function ($identifierName, $identifierValue) {
+            return sprintf('%s=%s', $identifierName, $this->generateIdentifiersUrl($identifierValue));
+        }, array_keys($identifiers), $identifiers);
```
Comment (soyuka) [inline src/Bridge/Symfony/Routing/IriConverter.php]: I'm not fond of this, does anyone have a better idea in mind?

**Comment by teohhanhui** [comment]
Can you explain what's going on here?

What did you mean when you said:
> This was working before #533

I don't see how that can be the case from the code before #533 - https://github.com/api-platform/core/blob/037fece1a681c6e747018ceef576a8b148b32910/src/Bridge/Symfony/Routing/IriConverter.php#L72-L90

We never checked for relations when getting identifier values.

**Comment by teohhanhui** [comment]
Anyway, for fixing this, why don't we just get the `id` from the entity if it's a relation? Like @Simperfit suggested...

**Comment by soyuka** [comment]
:confused: I can't say why it was working before, anyway maybe it wasn't at least it's now tested.

That's exactly what I did here `$identifiers[$propertyName] = $this->getIdentifiersFromItem($identifiers[$propertyName]);`, or you had an easier solution in mind?

**Comment by teohhanhui** [comment]
Your current method allows for recursion of more than 1 level. I'm not sure how we're going to actually make that work... For the case of related entity used as an identifier, we can only allow **non-composite identifier which is not a relation**.

**Comment by soyuka** [comment]
I can try to remove recursion, I like recursion too much I think :stuck_out_tongue_winking_eye:.

> For the case of related entity used as an identifier, we can only allow non-composite identifier which is not a relation.

Sorry I didn't understood. Isn't [this](https://github.com/api-platform/core/pull/552/commits/d387d916a63b190504712615d3317623dab62cad#diff-79bcda31abc0ceb00db05c7fee6b193cR72) enough? It lets the user choose if the identifier should be serialized or not, according to normalization groups.

**Comment by teohhanhui** [comment]
It has nothing to do with serialization. We're talking about the URL here...

**Comment by soyuka** [comment]
Sorry, though you were talking about the second commit. I rewrote 3bf5323, should be better now.

**Comment by Simperfit** [comment]
👍

**Comment by soyuka** [comment]
ping @api-platform/core lgtm

File: src/Bridge/Symfony/Routing/IriConverter.php [hunk 2]
```diff
@@ -77,25 +77,73 @@ public function getIriFromItem($item, int $referenceType = UrlGeneratorInterface
...
+
+            foreach ($this->propertyNameCollectionFactory->create($relatedResourceClass) as $relatedPropertyName) {
+                $propertyMetadata = $this->propertyMetadataFactory->create($relatedResourceClass, $relatedPropertyName);
+
+                if ($propertyMetadata->isIdentifier()) {
+                    $identifiers[$propertyName] = $this->propertyAccessor->getValue($relatedItem, $relatedPropertyName);
+                }
+            }
+
+            if (0 === count($identifiers[$propertyName])) {
```
Comment (dunglas) [inline src/Bridge/Symfony/Routing/IriConverter.php]: Prefer `empty` (for performance).

File: src/Bridge/Symfony/Routing/IriConverter.php [hunk 3]
```diff
@@ -77,25 +77,73 @@ public function getIriFromItem($item, int $referenceType = UrlGeneratorInterface
...
+     *
+     * @return array
+     */
+    public function generateIdentifiersUrl(array $identifiers) : array
+    {
+        if (1 === count($identifiers)) {
+            return [rawurlencode(array_values($identifiers)[0])];
+        }
+
+        return array_map(function ($identifierName, $identifierValue) {
```
Comment (dunglas) [inline src/Bridge/Symfony/Routing/IriConverter.php]: Why not a simple foreach (it is more performant than `array_map`)?

Comment (soyuka) [inline src/Bridge/Symfony/Routing/IriConverter.php]: js functional programming habits I guess...

**Comment by soyuka** [comment]
I can squash my commits but I think they are best separated, let me know.

**Comment by dunglas** [comment]
Thanks @soyuka!

**Pull request merged**